rotkehlchen/chain/ethereum/uniswap/uniswap.py

```diff
@@ -220,6 +242,197 @@ def _get_known_asset_price(
+                entry_name = f'{UNISWAP_TRADES_PREFIX}_{address}'
+                self.database.update_used_query_range(
+                    name=entry_name,
+                    start_ts=min_end_ts,
+                    end_ts=to_timestamp,
+                )
+
+        # Insert requested trades in DB
+        for address in filter(lambda address: address in address_amm_trades, addresses):
+            self.database.add_amm_trades(address_amm_trades[address])
+
+        # Fetch all DB Uniswap trades within the time range
+        for address in addresses:
+            db_trades = self.database.get_amm_trades(
+                from_ts=from_timestamp,
+                to_ts=to_timestamp,
+                location=Location.UNISWAP,
+                address=address,
+            )
+            if db_trades:
+                # return trades with most recent first
+                db_trades.sort(key=lambda trade: trade.timestamp, reverse=True)
+                db_address_trades[address] = db_trades
+
+        return db_address_trades
+
+    def _get_trades_graph(
+            self,
+            addresses: List[ChecksumEthAddress],
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> AddressTrades:
+        """Get the addresses' trades data querying the Uniswap subgraph
+
+        Each trade (swap) instantiates an <AMMTrade>.
+
+        The trade pair (i.e. BASE_QUOTE) is determined by `reserve0_reserve1`.
+        Translated to Uniswap lingo:
+
+        Trade type BUY:
+        - `asset1In` (QUOTE, reserve1) is gt 0.
+        - `asset0Out` (BASE, reserve0) is gt 0.
+
+        Trade type SELL:
+        - `asset0In` (BASE, reserve0) is gt 0.
+        - `asset1Out` (QUOTE, reserve1) is gt 0.
+        """
+        address_trades: DDAddressTrades = defaultdict(list)
+        addresses_lower = [address.lower() for address in addresses]
+        param_types = {
+            '$limit': 'Int!',
+            '$offset': 'Int!',
+            '$addresses': '[Bytes!]',
+            '$start_ts': 'BigInt!',
+            '$end_ts': 'BigInt!',
+        }
+        param_values = {
+            'limit': GRAPH_QUERY_LIMIT,
+            'offset': 0,
+            'addresses': addresses_lower,
+            'start_ts': str(start_ts),
+            'end_ts': str(end_ts),
+        }
+        querystr = format_query_indentation(SWAPS_QUERY.format())
+
+        while True:
+            result = self.graph.query(  # type: ignore # caller already checks
+                querystr=querystr,
+                param_types=param_types,
+                param_values=param_values,
+            )
+            result_data = result['swaps']
+
+            for swap in result_data:
+                user_address = to_checksum_address(swap['to'])
+                timestamp = swap['timestamp']
+                token0 = swap['pair']['token0']
+                token1 = swap['pair']['token1']
+                base_asset_rate = swap['pair']['token0Price']
```

Comment:
Using those to calculate the price is wrong. The`token0Price` and `token1Price` from pair is the **current** price of the token. Which naturally will not work with historical trades.

This is easy to notice from any example query. So this logic needs to change here.

Reply:
The fix is very easy and I am implementing it right now. You already got all the data you need to calculate a rate.